pennylane_lightning/src/util/LinearAlgebra.hpp

```diff
@@ -369,7 +369,7 @@ inline auto matrixVecProd(const std::vector<std::complex<T>> mat,
  * @param n1 Index of the first column.
  * @param n2 Index of the last column.
  */
-template <class T, size_t BLOCKSIZE = 32> // NOLINT(readability-magic-numbers)
+template <class T, size_t BLOCKSIZE = 16> // NOLINT(readability-magic-numbers)
```

Comment:
Does 16 offer better performance for this?

Reply:
Yes indeed! This is the result of running the following command:
```Bash
$ python3 compare.py filters ./BuildGBench/benchmarks/utils "cf_transpose_cmplx<double, 16>" "cf_transpose_cmplx<double, 32>"
```

```
Running ./BuildGBench/benchmarks/utils
Run on (8 X 3877.22 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
Load Average: 3.70, 2.74, 1.44
------------------------------------------------------------------------------
Benchmark                                    Time             CPU   Iterations
------------------------------------------------------------------------------
cf_transpose_cmplx<double, 16>/32          769 ns          769 ns       905299
cf_transpose_cmplx<double, 16>/64         3867 ns         3867 ns       181881
cf_transpose_cmplx<double, 16>/128       18741 ns        18741 ns        43778
cf_transpose_cmplx<double, 16>/256      223272 ns       223266 ns         3133
cf_transpose_cmplx<double, 16>/512     1028820 ns      1028753 ns          682
cf_transpose_cmplx<double, 16>/1024    5229414 ns      5229264 ns          129
cf_transpose_cmplx<double, 16>/2048   40673714 ns     40666706 ns           17
cf_transpose_cmplx<double, 16>/4096  165500143 ns    165467574 ns            4
cf_transpose_cmplx<double, 16>/8192  626944729 ns    626880717 ns            1
RUNNING: ./BuildGBench/benchmarks/utils --benchmark_filter=cf_transpose_cmplx<double, 32> --benchmark_out=/tmp/tmp_x_2ganq
2022-03-15T01:40:42-04:00
Running ./BuildGBench/benchmarks/utils
Run on (8 X 2341.99 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
Load Average: 3.10, 2.66, 1.43
------------------------------------------------------------------------------
Benchmark                                    Time             CPU   Iterations
------------------------------------------------------------------------------
cf_transpose_cmplx<double, 32>/32         1067 ns         1067 ns       670864
cf_transpose_cmplx<double, 32>/64         5940 ns         5940 ns       106912
cf_transpose_cmplx<double, 32>/128       23607 ns        23607 ns        29433
cf_transpose_cmplx<double, 32>/256      226348 ns       226345 ns         2899
cf_transpose_cmplx<double, 32>/512      999435 ns       999460 ns          668
cf_transpose_cmplx<double, 32>/1024    5251783 ns      5250924 ns          128
cf_transpose_cmplx<double, 32>/2048   39253467 ns     39204690 ns           17
cf_transpose_cmplx<double, 32>/4096  169543968 ns    169549855 ns            4
cf_transpose_cmplx<double, 32>/8192  639742739 ns    639680299 ns            1
Comparing cf_transpose_cmplx<double, 16> to cf_transpose_cmplx<double, 32> (from ./BuildGBench/benchmarks/utils)
Benchmark                                                                                  Time             CPU      Time Old      Time New       CPU Old       CPU New
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
[cf_transpose_cmplx<double, 16> vs. cf_transpose_cmplx<double, 32>]/32                  +0.3870         +0.3870           769          1067           769          1067
[cf_transpose_cmplx<double, 16> vs. cf_transpose_cmplx<double, 32>]/64                  +0.5363         +0.5363          3867          5940          3867          5940
[cf_transpose_cmplx<double, 16> vs. cf_transpose_cmplx<double, 32>]/128                 +0.2596         +0.2597         18741         23607         18741         23607
[cf_transpose_cmplx<double, 16> vs. cf_transpose_cmplx<double, 32>]/256                 +0.0138         +0.0138        223272        226348        223266        226345
[cf_transpose_cmplx<double, 16> vs. cf_transpose_cmplx<double, 32>]/512                 -0.0286         -0.0285       1028820        999435       1028753        999460
[cf_transpose_cmplx<double, 16> vs. cf_transpose_cmplx<double, 32>]/1024                +0.0043         +0.0041       5229414       5251783       5229264       5250924
[cf_transpose_cmplx<double, 16> vs. cf_transpose_cmplx<double, 32>]/2048                -0.0349         -0.0360      40673714      39253467      40666706      39204690
[cf_transpose_cmplx<double, 16> vs. cf_transpose_cmplx<double, 32>]/4096                +0.0244         +0.0247     165500143     169543968     165467574     169549855
[cf_transpose_cmplx<double, 16> vs. cf_transpose_cmplx<double, 32>]/8192                +0.0204         +0.0204     626944729     639742739     626880717     639680299
OVERALL_GEOMEAN                                                                         +0.1156         +0.1155             0             0             0             0
```

Reply:
This is a good idea. FYI, `BLOCKSIZE=2^n` performs transposition over submatrices of size `2^n * 2^n` of the original matrix and the performance we gain using this blocking technique should come from the size of cache and the number of cache misses. On my machines with the following cache info, transposing submatrices of size `2^8` is more cache-friendly than transposing submatrices of size `2^10`,
```console
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
```